src/consensus/aft/raft.h

```diff
@@ -1259,45 +1284,98 @@ namespace aft
       }
       else
       {
-        guard.unlock();
-        msg->cb(std::move(msg));
+        apply_execution_message(std::move(msg));
       }
     }
 
+    struct AsyncExecutionRet
+    {
+      AsyncExecutionRet(
+        Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self_) :
+        self(self_)
+      {}
+
+      Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self;
+    };
+
     static void execute_append_entries_cb(
       std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
     {
-      msg->data.self->execute_append_entries(
-        msg->data.append_entries, msg->data.r, msg->data.confirm_evidence);
-      msg->cb =
-        reinterpret_cast<void (*)(std::unique_ptr<threading::ThreadMsg>)>(
-          continue_execution);
+      auto self = msg->data.self;
+      std::unique_lock<SpinLock> guard(self->state->lock);
+      self->apply_execution_message(std::move(msg));
+    }
+
+    void apply_execution_message(
+      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+    {
+      auto self = msg->data.self;
+      if (self->consensus_type == ConsensusType::CFT)
+      {
+        self->execute_append_entries_sync(msg);
+      }
+      else
+      {
+        if (
+          self->execute_append_entries_async(msg) ==
+          AsyncExecutionResult::PENDING)
+        {
+          return;
+        }
+      }
+
+      auto msg_ret = std::make_unique<threading::Tmsg<AsyncExecutionRet>>(
+        continue_execution, self);
       if (threading::ThreadMessaging::thread_count > 1)
       {
         threading::ThreadMessaging::thread_messaging.add_task(
-          threading::MAIN_THREAD_ID, std::move(msg));
+          threading::MAIN_THREAD_ID, std::move(msg_ret));
       }
       else
       {
-        msg->cb(std::move(msg));
+        msg_ret->cb(std::move(msg_ret));
       }
     }
 
     static void continue_execution(
-      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+      std::unique_ptr<threading::Tmsg<AsyncExecutionRet>> msg)
     {
       msg->data.self->is_execution_pending = false;
       msg->data.self->try_execute_pending();
     }
 
-    void execute_append_entries(
+    struct AsyncExecTxMsg
+    {
+      AsyncExecTxMsg(
+        Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self_,
+        std::unique_ptr<kv::AbstractExecutionWrapper>&& ds_,
+        kv::Version last_idx_,
+        kv::Version commit_idx_,
+        uint16_t scheduler_thread_) :
+        self(self_),
+        ds(std::move(ds_)),
+        last_idx(last_idx_),
+        commit_idx(commit_idx_),
+        scheduler_thread(scheduler_thread_)
+      {}
+
+      Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self;
+      std::unique_ptr<kv::AbstractExecutionWrapper> ds;
+      kv::Version last_idx;
+      kv::Version commit_idx;
+      uint16_t scheduler_thread;
+      std::shared_ptr<AsyncExecutor> ctx;
+    };
+
+    void execute_append_entries_sync(
```

Comment:
comment -> we have a dupe code path to derisk for 1.0 should look at combining after 1.0